netlink.go

```diff
@@ -3,25 +3,195 @@
+		fmt.Printf("\r\n")
+	}
+	d.deviceShown = true
+}
+
+func (d *Device) connectToAP() error {
+
+	if len(d.params.Ssid) == 0 {
+		return netlink.ErrMissingSSID
+	}
+
+	if debugging(debugBasic) {
+		fmt.Printf("Connecting to Wifi SSID '%s'...", d.params.Ssid)
+	}
+
+	err := d.wifiConnectTimeout()
+	if err != nil {
+		if debugging(debugBasic) {
+			fmt.Printf("FAILED (%s)\r\n", err.Error())
+		}
+		return err
+	}
+
+	if debugging(debugBasic) {
+		fmt.Printf("CONNECTED\r\n")
+	}
+
+	if d.notifyCb != nil {
+		d.notifyCb(netlink.EventNetUp)
+	}
+
+	return nil
+}
+
+func (d *Device) networkDown() bool {
+	return false
+}
+
+func (d *Device) showIP() {
+}
+
+func (d *Device) netConnect(reset bool) error {
+	if reset {
+		if err := d.enableStaMode(); err != nil {
+			return err
+		}
+	}
+	d.showDevice()
+
+	for i := 0; d.params.Retries == 0 || i < d.params.Retries; i++ {
+		if err := d.connectToAP(); err != nil {
+			if err == netlink.ErrConnectFailed {
+				continue
+			}
+			return err
+		}
+		break
+	}
+
+	if d.networkDown() {
+		return netlink.ErrConnectFailed
+	}
+
+	d.showIP()
+	return nil
+}
+
+func (d *Device) watchdog() {
+	ticker := time.NewTicker(d.params.WatchdogTimeout)
+	for {
+		select {
+		case <-d.killWatchdog:
+			return
+		case <-ticker.C:
+			d.mu.Lock()
+			if d.networkDown() {
+				if debugging(debugBasic) {
+					fmt.Printf("Watchdog: Wifi NOT CONNECTED, trying again...\r\n")
+				}
+				if d.notifyCb != nil {
+					d.notifyCb(netlink.EventNetDown)
+				}
+				d.netConnect(false)
+			}
+			d.mu.Unlock()
+		}
+	}
+}
+
+func (d *Device) NetConnect(params *netlink.ConnectParams) error {
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.netConnected {
+		return netlink.ErrConnected
+	}
+
+	switch params.ConnectMode {
+	case netlink.ConnectModeSTA:
+	default:
+		return netlink.ErrConnectModeNoGood
+	}
+
+	d.params = params
+
+	d.showDriver()
+
+	if err := d.netConnect(true); err != nil {
+		return err
+	}
+
+	d.netConnected = true
+
+	if d.params.WatchdogTimeout != 0 {
+		go d.watchdog()
+	}
+
+	return nil
+}
+
+func (d *Device) netDisconnect() {
 }
 
 func (d *Device) NetDisconnect() {
+
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if !d.netConnected {
+		return
+	}
+
+	if d.params.WatchdogTimeout != 0 {
+		d.killWatchdog <- true
+	}
+
+	d.netDisconnect()
+	d.netConnected = false
+
+	if debugging(debugBasic) {
+		fmt.Printf("\r\nDisconnected from Wifi\r\n\r\n")
+	}
+
+	if d.notifyCb != nil {
+		d.notifyCb(netlink.EventNetDown)
+	}
 }
 
-func (d *Device) NetNotify(cb func(drivers.NetlinkEvent)) {
+func (d *Device) NetNotify(cb func(netlink.Event)) {
+	d.notifyCb = cb
 }
 
 func (d *Device) GetHardwareAddr() (net.HardwareAddr, error) {
-	return net.HardwareAddr{}, drivers.ErrNotSupported
+	return d.mac, nil
+}
+
+func (d *Device) SendEth(pkt []byte) error {
+	return netlink.ErrNotSupported
 }
 
-func (d *Device) GetIPAddr() (net.IP, error) {
-	return net.IP{}, drivers.ErrNotSupported
+func (d *Device) RecvEthFunc(cb func(pkt []byte) error) {
```

Comment:
Probably call this SetEthPacketRxCallback, or the likes

Reply:
RecvEthCallback is fine

Reply:
Hmmm...SetRecvEthCallback sounded good, but then I started thinking about "Callback".  (I know I used "cb" for arg also).  It's not really a callback; it's more the Rx handler func.  Would SetRecvEthHandler be a better name?  Or taking inspiration from the net/http package, do something like RecvEthHandle(handler func(...))?  They seem to drop the "Set" from the name.

So we'd have SendEth() and RecvEthHandle().  Looking for some symmetry here also, between send and recv.

I don't know, naming it hard.  Which would you lean towards?

Reply:
Oooo nice handler sounds nice. Go for it

Reply:
I think my ranked choices are: RecvEthHandle, RecvEthFunc, RecvEthCallback, followed by the SetXxx variants.

Reply:
RecvEthHandle's good :+1: